### ISSUE TITLE
Add an option for file nesting in the FileSystem dock

### ISSUE THREAD

**Issue by pycbouh** [godotengine/godot-proposals#3798]
### Describe the project you are working on

Godot Editor usability

### Describe the problem or limitation you are having in your project

Project organization is an important aspect of keeping your sanity when working with mid- and large-scale projects. One of the problems that comes up often enough in various proposals is whether having scripts next to scenes is creating a lot of clutter in the filesystem dock or not. It's also often stated as a good reason to use built-in scripts.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I think we can offer a new option for the filesystem dock in the form of "file nesting" feature. If you've ever used JetBrains' IDEs you've probably seen how sometimes several related files are "collapsed" into one entry in its project outliner with the most important file being the root, and the axillary files being under the fold. For example, you can see this here with a TypeScript script being the root and its JS output as well as transpilation mapping file being folded.

![image](https://user-images.githubusercontent.com/11782833/149537506-f9232966-b66c-452c-8689-560abc1f29f9.png)
*Image from [WebStorm documentation](https://www.jetbrains.com/help/webstorm/file-nesting-dialog.html).*

I think a similar approach can be used to collapse pairs of `MyScene.tscn` and `MyScene.gd` (and other scripts) located in the same directory. It can be done by the name only without checking if the script is actually attached to the scene, because the general assumption would be that it is in this case.

Additionally, the same folded look can be used to give quick access to built-in scripts of a scene. Normally, they are not visible in the filesystem dock, and for a good reason, but we can expose another option here to display them and do so in this file nested form.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For the scene/script file pairs it should be trivial to change how the filesystem tree is constructed to achieve the look. It's just a `Tree` control and a script file can easily be made a child of a scene file. Clicking on each file should open it, like before, but the scene one will also have a collapse icon now, which should not trigger file opening.

It can be a bit more difficult to do the second suggestion regarding the built-in scripts, as it requires the knowledge of the contents of the scene. I'm not entirely sure, if this information is readily available without having to parse each scene. It should be, but worst-case scenario, filesystem dock doesn't update too often to do the parsing.

I'll investigate the best solution in the future and PR it.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is an improvement to the editor UX. It doesn't provide any features that are not already available, but it should improve their convenience.

### Is there a reason why this should be core and not an add-on in the asset library?

This is an improvement to the editor UX.

**Comment by SilencedPerson** [comment]
Interesting proposal.
So this works by having the everything after the dot ignored and bundled under the scene? 
Let's say I have MainScene.tscn, it's script, MainScene.gd then becomes bundled, what if I make another script called MainScene.ChildNode.gd, would it also be bundled under the MainScene.tscn? If so, could it ignore slashes too?

Main reason why I use Built-In scripts is, because I don't care how the scripts are named and where they are stored. Built in scripts are very convenient because I just have to click one button and the engine does this for me and as a nice bonus, the filesystem doesn't get cluttered. I would probably didn't rely on built in scripts as heavily, if these two things were handled by the editor. Simply using the whole path in the SceneTree as a base name and simply bundling them under the scene in which they are contained would be sufficient for me.

**Comment by pycbouh** [comment]
> what if I make another script called MainScene.ChildNode.gd, would it also be nested under the MainScene.tscn?

I didn't consider something like that because that's not a typical naming convention for Godot projects. This is of course absolutely possible to have, technically, but I'm not sure if it should be done as a part of the same option, or if it needs a separate option, or if we should just allow users defining a mask (probably not, though)

Besides, if both `MainScene.gd` and `MainScene.ChildNode.gd` were somehow collapsed under the scene file, it wouldn't really save you a click. To access the second script file you'd still have to unfold the group.

-----
But for built-in scripts I want to have all of them listed, as they are all definitely associated with the scene.

**Comment by SilencedPerson** [comment]
In this case, I do not care about saving a click when accessing a node, rather, I am thinking about not having to type a name of a script when I create new one, saving a click when I am saving it to a filesystem, saving a click when I create resource, saving a click when I create another script file in the same scene and so on, not having to create folders. It can potentially save whole lot of clicks and typing. If this was implemented, it would immediately change how the users would treat the filesystem.

For example, There would probably be very little use for folders, User could just save a scene with unique name, then the Editor would Nest everything automatically.

**Comment by SilencedPerson** [comment]
For example, just by using unique names for scene files, (which is common anyway):
Player.tscn
Player.Sprites.WalkAnimations.Left.png
Player.Sprites.WalkAnimations.Right.png
Player.Sprites.WalkAnimations.Idle.png
Player.gd
Player.Camera.gd

would become:
Player >

**Comment by SilencedPerson** [comment]
Additionally, having a folder named Player.Sprites would then allow having said Folder Nested under the Player.tscn
and saving the sprites into that folder.

**Comment by spindlebink** [comment]
My issue with this proposal is that it would elevate a particular type of file structure over another. The IDEs that do this (AFAIK) do so for files that are, based on the tools they're written for, intrinsically linked--TypeScript compiles to JS; certain UI frameworks demand a certain file structure, etc. The IDE makes that connection explicit. It's more convenient and readable that way.

Godot scenes & scripts don't work like that. Scripts exist independently of their scenes and may be attached to multiple scenes, etc. Different people use radically different ways of structuring their programs (e.g. I've seen a lot of people use `scripts/*.gd`/`scenes/*.tscn`). And combining them like this in the editor would be giving implicit precedence to one variety of organization over another and communicating a semantic relationship between components that really doesn't exist.

Maybe it'd be better to nest *attached* scripts or resources under the scene if they're in the same directory? Or provide a "all associated resources" dropdown that could be accessed in the file browser? But the former case seems like a decent amount of work for a relatively small gain and the latter doesn't address the core issue of organization/visual hierarchy for files. Most extreme of course would be to make this possible relationship explicit and enforce structure like this, but I feel like that's counter to Godot's goals and not a great idea.

**Comment by SilencedPerson** [comment]
There is nothing wrong with having favored file structure and nothing would really change for the user if they did not want to use this system and the directories would still be there and taken into account, i can't see those going away under any circumstance.

If user preferred to have `scripts/*.gd` and `scenes/*.tscn`, then they can still do that and the nesting would only come into account if they had scene with a same script name which would not happen in their case, also in general, scripts that are shared between multiple scenes, tend to have different, more general name so those would not end up being nested either.

For example:
```
Scripts
    ┖ ExampleScene.gd
    ┖ ScriptSharedBetweenMultipleScenes.gd
Scenes
    ┖ ExampleScene.tscn
ScenesWithScripts
    ┖ ExampleScene.tscn
    ┖ Scene_1.gd
```
Would not result in any nesting.

Nesting would happen only if there were two files, one `*.gd` and one `*.tscn` file in the same directory.

Please do not take my comments and exploration of the concept earlier into the account, that's me taking things to extreme because that's how I would immediately change the way I structure my project to conform and use this feature to it's maximum potential.

**Comment by spindlebink** [comment]
No, you're right that there's nothing inherently wrong about having a preferred structure, I guess my question is just whether this structure is actually preferred by Godot's core team/style guide/whatever. If it is, then adopting this proposal would mean Godot would need to change the language used to describe scene organization/structure to prefer this one and do a soft deprecation of other styles.

What I'm saying is that this proposal is more than just adding a feature to visually nest files--it requires asking that question about preferred scene/script structure, coming to a decision (this'd be for the Godot core team to decide) and employing that engine-wide. Otherwise I think it'd feel like a weird quirk--*oh, and if you happen to organize your scenes like this, Godot has a feature where it nests your files and organizes them for you, but you don't have to organize them like this, and if you don't, you won't even know this feature exists*.

Also, bigger I think than that concern is the implied semantic relationship between files involved here. Again, nesting files like this communicates something about their relationship that may not be present. I can already imagine a cascade of beginner questions in the Discord about why scripts aren't taking effect on scenes when Godot clearly picks them up as connected, or losing track of which scripts and scenes are connected, etc. You can't really underestimate the confusion that can be caused by visuals and semantics not lining up, even if it's so minor as nesting files that're named similarly.

I'm not trying to be a naysayer, and since I personally do use this file structure, I really like the idea for my own projects. I guess I just see it as going counter to the principle of least surprise, at least as a core engine feature. I haven't really written any editor addons, but maybe that'd be an option. Alternatively, it's possible this could become a project setting, possibly with more customizability for nesting rules, or even a non-judgmental version that allows you to drop one file on another to create your own nesting hierarchy and store that info in the project metadata?

**Comment by pycbouh** [comment]
> No, you're right that there's nothing inherently wrong about having a preferred structure, I guess my question is just whether this structure is actually preferred by Godot's core team/style guide/whatever

It is, in fact. Putting related files next to each other is what we recommend in the official project organization guide in the docs. And when you attach a script to a node, the default behavior is to put it next to the scene, and give it the same name as the scene. So for most people, it's either preferred or simply default behavior.

But that's not really a deal breaker if you want to rebel against official guidelines. This proposal only adds an option, you don't have to use it even if you follow the guidelines. It can be to your liking or not. It will be opt-in, so you decide it for yourself. I think you are looking too deep into philosophy of this simple change.

If I can help it, I would add a check that the script is actually attached to the root of the scene. But that may be harder and more resource intensive than it needs to be. So the assumption that a scene and a script file of the same base name next to each other are in fact connected is not that far fetched. That's the most natural state Godot projects have, in fact.

**Comment by spindlebink** [comment]
That makes a lot of sense then, if it is in fact the recommended way. I may've been spending too much time lurking in the Rust community, where every suggestion has to be hashed and rehashed to determine whether it's adequately divinely-inspired or not.

**Comment by winston-yallow** [comment]
I think we can break this into multiple aspects:

### File-Association
Which scripts are shown as associated? I think there basically are two ways to go about this:
- Name determines association: The files are associated when the name matches. This could potentially work on all files, not only scripts. So all of `Player.{tscn,gd,png}` could be shown under the `Player` item. I also know that some people use PascalCase for scenes and snake_case for scripts. I think we should do automatic matching between those (probably as an option).
- Usage determines association: Name is irrelevant, instead the tscn is parsed and all scripts and resources (which are in the same filesystem folder) are associated with that node

I personally prefer the name based association. Additionally, all built-in scripts should be shown. The name based association also prevents one file to be associated to multiple scenes. Otherwise two scenes using the same script would both show that script. I personally think the name based approach is more transparent to the user.

### Quick Access
How to still quickly access stuff? I really like the idea to have script icons directly in the file system dock. This sounds like a good approach.
Additionally, I think that opening the `Player` item (which has the `{tscn,gd,png}` childs) should open *all* files that are associated in the appropiate editors (that is, the scene as a scene tab, scipts in the script editor, shaders in the shader editor, materials in the inspector and so on). If multiple files would be opened in an editor, it should use the one that is first. The other ones should also be opened if the editor supports multiple files be openend (like multiple scripts, but the first one will be the currently edited one). I think this could really speed up my workflow, as I often double click all associated files by hand when starting to edit a scene.

### Layout
How should it look? I personally would like a completely new icon (not a folder icon) for the group item (like `Player` in this example). It should communicate that it is some kind of virtual collection, not a real folder. I personally would like something like the "Category" or "Apps" icon from here: https://fonts.google.com/icons?icon.query=collection
If my idea from the Quick Access section is wanted, then I suggest having a top level item that has all associated files (including the `tscn` itself) as children. So a click on the top level item opens all the stuff, but you can still open only the scene by opening that one. If we don't want that idea, then I think having the `tscn` as the top level is better, as it communicates clearly that it will only open that. It is not needed as a child in that case.

---

Ok, so the above are ideas that are kinda close to how this was proposed in the first place. But I have a completely separate but related idea too:
Keep the filesystem dock as a pure filesystem. Add an additional Dock "Project Outline". It shows stuff by association. It would have a section "Scenes", and each scene would have all it's associated stuff as children. Could have other sections than "Scenes" too, like "Assets" to browser assets by type and so on. Such a Dock would not be declared as a file system and can allow for stuff to show up in multiple places without it becoming confusing. It is meant as a quick way to explore the project, not as a way to see where stuff is stored.
I have not thought about any details of this approach or if it is feasable at all. I could come up with more detailed approaches to this, but I first wanted to see if this direction is wanted at all. I am myself not sure if this is the way to go. And even if this is wanted, it could still be helpful to have it *additionally* to this Proposal. Maybe even as an addon since it just adds a new dock?

**Comment by pycbouh** [comment]
> Ok, so the above are ideas that are kinda close to how this was proposed in the first place. But I have a completely separate but related idea too:
Keep the filesystem dock as a pure filesystem. Add an additional Dock "Project Outline". It shows stuff by association. It would have a section "Scenes", and each scene would have all it's associated stuff as children. Could have other sections than "Scenes" too, like "Assets" to browse assets by type and so on. Such a Dock would not be declared as a file system and can allow for stuff to show up in multiple places without it becoming confusing. It is meant as a quick way to explore the project, not as a way to see where stuff is stored.
I have not thought about any details of this approach or if it is feasable at all. I could come up with more detailed ideas to this, but I first wanted to see if this direction is wanted at all. I am myself not sure if this is the way to go. And even if this is wanted, it could still be helpful to have it additionally to this Proposal. Maybe even as an addon since it just adds a new dock?

@winston-yallow Not entirely what you describe, but how about this https://github.com/pycbouh/godot-editor-assets-dock ? 👀

**Comment by winston-yallow** [comment]
Yes, something like that, but it would need a way to quickly have a list of associated stuff of a scene. Let's see where this Proposal goes, maybe a combination of this proposal and your addon will turn out to be perfect for my workflow ✨

**Comment by me2beats** [comment]
in addons folder: [plugin_name].gd could be on top.

**Comment by winston-yallow** [comment]
I don't think changing the order of files in some directories is a good idea. (Beside that, the main file of a plugin isn't always named like the plugin).

I think it would be more confusing than useful. If the addon has many files, one would use subfolders for them anyway. So rearranging the file order wouldn't really make it easier to find the plugin main script, it would just make the ordering a bit more confusing.

Also this proposal is about nesting the files. I'm not sure if reordering would be within the scope of this proposal. It probably would need it's own proposal.
